app/api/views/teams.py

```diff
@@ -120,3 +120,84 @@ def request_access():
         abort(e.message)
 
     return jsonify(success=True)
+
+
+@api.route('/team/<int:team_id>/request-join', methods=['POST'])
+@exception_logger
+@login_required
+@role_required('buyer')
+def request_to_join(team_id):
+    try:
+        team_business.request_to_join(current_user.email_address, team_id, current_user.agency_id)
+    except NotFoundError as e:
+        return not_found(e.message)
+    except UnauthorisedError as e:
+        return forbidden(e.message)
+    except ValidationError as e:
+        abort(e.message)
+
+    return jsonify(success=True)
+
+
+@api.route('/team/join-requests', methods=['GET'])
+@exception_logger
+@login_required
+@role_required('buyer')
+def get_join_requests():
+    join_requests = team_business.get_join_requests(current_user.email_address)
+    results = {}
+    if join_requests:
+        for request in join_requests:
+            if request.data['team_id'] not in results:
+                results[request.data['team_id']] = []
+            results[request.data['team_id']].append({
+                "id": request.id,
+                "team_id": request.data['team_id'],
+                "created_at": request.created_at
+            })
+    return jsonify(join_requests=results)
+
+
+@api.route('/team/join-request/<int:team_id>/<string:token>', methods=['GET'])
+@exception_logger
+@login_required
+@role_required('buyer')
+def get_join_request(team_id, token):
+    try:
+        team = team_business.get_team(team_id)
+    except NotFoundError as e:
+        return not_found(e.message)
+    except UnauthorisedError as e:
+        return forbidden(e.message)
+
+    join_request = team_business.get_join_request(token)
+    if not join_request or int(join_request.data['team_id']) != team_id:
+        return not_found('Invalid token')
+
+    return jsonify(join_request=join_request)
+
+
+@api.route('/team/decline-join-request/<int:team_id>/<string:token>', methods=['POST'])
+@exception_logger
+@login_required
+@role_required('buyer')
+def decline_join_request(team_id, token):
+    data = get_json_from_request()
+
+    try:
+        team = team_business.get_team(team_id)
+    except NotFoundError as e:
+        return not_found(e.message)
+    except UnauthorisedError as e:
+        return forbidden(e.message)
+
+    if 'reason' not in data or not data['reason']:
```

Comment:
shouldn't this check happen inside: team_business.decline_join_request(join_request, data['reason'])

Reply:
I don't think so - the `reason` arg in `decline_join_request()` is not optional, so its presence in the request needs to be checked, and there is no need to do the `get_join_request()` call if the request is invalid so the earlier it is checked the better (and in fact I'll move it before `get_team()` as it doesn't need to do that if the reason isn't provided).